src/linker/Linker.Steps/UnreachableBlocksOptimizer.cs

```diff
@@ -1332,80 +1361,58 @@ void CleanExceptionHandlers ()
 
 				return null;
 			}
-
-			static bool IsSideEffectFreeLoad (Instruction instr)
-			{
-				switch (instr.OpCode.Code) {
-				case Code.Ldarg:
-				case Code.Ldloc:
-				case Code.Ldloc_0:
-				case Code.Ldloc_1:
-				case Code.Ldloc_2:
-				case Code.Ldloc_3:
-				case Code.Ldloc_S:
-				case Code.Ldc_I4_0:
-				case Code.Ldc_I4_1:
-				case Code.Ldc_I4_2:
-				case Code.Ldc_I4_3:
-				case Code.Ldc_I4_4:
-				case Code.Ldc_I4_5:
-				case Code.Ldc_I4_6:
-				case Code.Ldc_I4_7:
-				case Code.Ldc_I4_8:
-				case Code.Ldc_I4:
-				case Code.Ldc_I4_S:
-				case Code.Ldc_I4_M1:
-				case Code.Ldc_I8:
-				case Code.Ldc_R4:
-				case Code.Ldc_R8:
-				case Code.Ldnull:
-				case Code.Ldstr:
-					return true;
-				}
-
-				return false;
-			}
 		}
 
 		struct ConstantExpressionMethodAnalyzer
 		{
-			readonly MethodDefinition method;
-			readonly Collection<Instruction> instructions;
 			readonly LinkContext context;
+			readonly UnreachableBlocksOptimizer optimizer;
 
 			Stack<Instruction>? stack_instr;
 			Dictionary<int, Instruction>? locals;
 
-			public ConstantExpressionMethodAnalyzer (LinkContext context, MethodDefinition method)
+			public ConstantExpressionMethodAnalyzer (UnreachableBlocksOptimizer optimizer)
 			{
-				this.context = context;
-				this.method = method;
-				instructions = method.Body.Instructions;
+				this.optimizer = optimizer;
+				this.context = optimizer._context;
 				stack_instr = null;
 				locals = null;
 				Result = null;
+				SideEffectFreeResult = true;
 			}
 
-			public ConstantExpressionMethodAnalyzer (LinkContext context, MethodDefinition method, Collection<Instruction> instructions)
-				: this (context, method)
-			{
-				this.instructions = instructions;
-			}
-
+			//
+			// Single expression that is representing the evaluation result with the specific
+			// callee arguments
+			//
 			public Instruction? Result { get; private set; }
 
+			//
+			// Returns true when the method evaluation with specific arguments does not cause
+			// any observable side effect (e.g. possible NRE, field access, etc)
+			//
+			public bool SideEffectFreeResult { get; private set; }
+
 			[MemberNotNullWhen (true, "Result")]
-			public bool Analyze ()
+			public bool Analyze (in CalleePayload callee, Stack<MethodDefinition> callStack)
```

Comment:
This is very similar to MethodBodyScanner - might be worth looking into potential sharing. Or at the very least comparing the two implementations.

Reply:
That said - with the linker switching over to the "solver based data flow", the method body scanner will probably change quite a bit - especially around branching and such (basically it will become BasicBlockScanner instead).